gravitee-io/gravitee-api-management
# feat: debug mode - display where a policy is configured

**Issue**

https://github.com/gravitee-io/issues/issues/7183

**Description**

1. Externalize PolicyResolver.Policy inner class to have a dedicated class called `PolicyMetadata`. Existing `PolicyMetadata` class is renamed to `PolicyManifest` because its purpose is to build the policy.
2. Instead of passing parameters to build a `Policy` thanks to `PolicyManager.create()`, we now pass the `PolicyMetadata` object which holds the policy condition too.
3. Add the new `stage` info into `PolicyMetadata`. This stage is defined at the `FlowResolver` level. 
4. `*FlowChainprovider` needed a `SimpleFlowProvider` to be created. It makes it hard to read the class without providing that much value. We now delegate the creation of this `SimpleFlowProvider` into the constructor of `*FlowChainprovider` 
5. Improve performance avoiding creating a `FlowPolicyResolver` for each `Flow`. Same thing is done to avoid instantiating `PolicyMetadata` for each step of a flow.
6. Manage stage for Security level policies. Also, manage the case of internal policies (`CheckSubscription` policy is executed for Oauth2 or JWT plans for example.)
7. Improve display of card in Debug Mode when they have no icon to display. 
8. Display policy stage in cards

## To test

Test the Debug Mode as usual paying attention at the stage.
Also, test that "Check Subscription" policy appears for plans like Oauth2 or JWT.

![image](https://user-images.githubusercontent.com/47851994/157876786-97c4e389-ec47-419c-bcab-9b11d23ea125.png)


📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-nswthuwqkl.chromatic.com)


---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-zbjidjcrkh.chromatic.com)


🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.blob.core.windows.net/7183-where-policy-v2/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)
